app/src/main/java/com/example/hlupico/mememaker/NewMemeActivity.java

```diff
@@ -1,31 +1,167 @@
+            File newfile = new File(getExternalFilesDir(Environment.DIRECTORY_PICTURES).getPath(), imageFileName);
+            try {
+                newfile.createNewFile();
+            } catch (IOException e) {}
+
+            tempImageUri = FileProvider.getUriForFile(
+                    this,
+                    getApplicationContext()
+                            .getPackageName() + ".provider", newfile);
+
+            Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
+            takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, tempImageUri);
+            if (takePictureIntent.resolveActivity(getPackageManager()) != null) { // device has a camera app
+                startActivityForResult(takePictureIntent, REQUEST_CODE_TAKE_PHOTO);
+            }
+        }
+    }
+
+    private void openGallery() {
+        boolean storageReadPermissionGranted = ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) ==
+                PackageManager.PERMISSION_GRANTED;
+        if (storageReadPermissionGranted == false) {
+            ActivityCompat.requestPermissions(this, GALLERY_PERMISSION, REQUEST_CODE_GALLERY);
+        } else {
+            Intent intent = new Intent();
+            // Show only images, no videos or anything else
+            intent.setType("image/*");
+            intent.setAction(Intent.ACTION_GET_CONTENT);
+            // Always show the chooser (if there are multiple options available)
+            startActivityForResult(Intent.createChooser(intent, "Select Picture"), REQUEST_CODE_CHOOSE_PHOTO);
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        if (requestCode == REQUEST_CODE_CAMERA) {
+            // did the user give us permission?
+            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                openCamera();
+            } else {
+                Toast.makeText(this, "Need permissions to take a photo", Toast.LENGTH_LONG).show();
+            }
+        } else if (requestCode == REQUEST_CODE_GALLERY) {
+            // did the user give us permission?
+            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                openGallery();
+            } else {
+                Toast.makeText(this, "Need permissions to select a photo", Toast.LENGTH_LONG).show();
+            }
+        }
+    }
+
+    private void setThumbnail(Bitmap image) {
+        ImageView thumbnail = (ImageView)findViewById(R.id.image_thumbnail);
+        thumbnail.setImageBitmap(image);
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == REQUEST_CODE_TAKE_PHOTO && resultCode == RESULT_OK) {
+            imageUri = tempImageUri;
+            try {
+                memeImage = MediaStore.Images.Media.getBitmap(getContentResolver(), imageUri);
```

Comment:
Could we also save the photo file at this time? If you take a photo, go back to `Make New Meme` without following through with Preview -> Save, the photo will be lost. 

If we save the photo here it will be in the Gallery if  they wanted to retrieve it later or if they wanted to use the same picture for multiple memes.